### PR TITLE
[REEF-1822] Create a script to analyze CLR Compliance

### PR DESCRIPTION
**Pull request by rogancarr**
This addressed the issue by
  * Implementing a script that analyzes the build artifacts for compliance of the various CLR runtimes

JIRA:
  [REEF-1822](https://issues.apache.org/jira/browse/REEF-1822)

### CODE REVIEW

**Comment by rogancarr** [comment]
This is my first ever PowerShell script, so please let me know if there are coding standards or conventions that I'm missing here.

File: lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1 [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+		{
+			echo "Skipping ${projectName}: No csproj"
+			return
+		}
+
+		# Determine the output type for the project
+		$outputType = $(Select-String -Path $csProj -Pattern "<OutputType>").Line -replace '^\s+<OutputType>(.+?)</OutputType>\s*$','$1'
+
+		# Get the portability settings for this project
+		$objectSuffix = ""
```
Comment (rogancarr) [inline lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1]: I assume that each project creates one final package -- a dll or an exe.

File: lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1 [hunk 2]
```diff
@@ -0,0 +1,108 @@
+# Input / Output Settings
+$baseDir = "..\"
+$binDir="${baseDir}\bin\x64\Debug"
+$outputDirectory = "Output"
+
+# Compatibility Analyzer settings
+$libraryDotNetTargets = @(".NET Standard, Version=2.0")
+$exeDotNetTargets = @(".NET Core + Platform Extensions, Version=1.0", ".NET Framework, Version=4.5.1")
```
Comment (rogancarr) [inline lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1]: From @markusweimer 's comments on the dev list. I'm assuming that we want 4.5.1 and NOT 4.6.1.

File: lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1 [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	[IO.Compression.Zipfile]::ExtractToDirectory($analyzerZip, (pwd));
+	rm $analyzerZip
+}
+
+# Set up the counter
+[int] $totalIncompatibilities = 0
+
+# Start a unique list of dependencies
+$missingDependencies = @{}
+
```
Comment (rogancarr) [inline lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1]: Any other factoids we want to report?

**Comment by rogancarr** [comment]
This is what the output will look like: (run on master 70ca4d92c118accf51bb73c69d94cf8a8878a614)

PS C:\Users\rogan\Projects\reef\lang\cs\CoreCLRCompatibility> .\AnalyzeCoreCLRCompatibility.ps1
Analyzing Org.Apache.REEF.All - Library
        Found 0 incompatibilities
Skipping Org.Apache.REEF.Bridge: No csproj
Skipping Org.Apache.REEF.Bridge.JAR: OutputType is not 'Library','Exe'
Analyzing Org.Apache.REEF.Client - Library
        Found 2 incompatibilities
Analyzing Org.Apache.REEF.Client.Tests - Library
        Found 0 incompatibilities
Skipping Org.Apache.REEF.ClrDriver: No csproj
Analyzing Org.Apache.REEF.Common - Library
        Found 12 incompatibilities
Analyzing Org.Apache.REEF.Common.Tests - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Driver - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Evaluator - Exe
        Found 1 incompatibilities
Analyzing Org.Apache.REEF.Evaluator.Tests - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Examples - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Examples.AllHandlers - Exe
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Examples.DriverRestart - Exe
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Examples.HelloREEF - Exe
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.FatNuGet - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.IMRU - Library
        Found 1 incompatibilities
Analyzing Org.Apache.REEF.IMRU.Examples - Exe
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.IMRU.Tests - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.IO - Library
        Found 6 incompatibilities
Analyzing Org.Apache.REEF.IO.TestClient - Exe
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.IO.Tests - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Network - Library
        Found 12 incompatibilities
Analyzing Org.Apache.REEF.Network.Examples - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Network.Examples.Client - Exe
        Found 1 incompatibilities
Analyzing Org.Apache.REEF.Network.Tests - Library
        Found 2 incompatibilities
Analyzing Org.Apache.REEF.Tang - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Tang.Examples - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Tang.Tests - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Tang.Tools - Exe
        Found 1 incompatibilities
Analyzing Org.Apache.REEF.Tests - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Utilities - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Wake - Library
        Found 0 incompatibilities
Analyzing Org.Apache.REEF.Wake.Tests - Library
        Found 2 incompatibilities
Found 40 total incompatibilities
Missing Dependencies:
        mscorlib
        System.Runtime.Caching
        System
        System.Reactive.Core
        System.Net.Http.WebRequest

**Comment by markusweimer** [comment]
This looks great! Any chance we can store the number of incompatibilities in a file and then fail the build if the number goes up?

File: lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1 [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+# Install the API Portability Analyzer if it is not present
+$portabilityAnalyzer = ".\ApiPort\ApiPort.exe"
+If (!(Test-Path $portabilityAnalyzer))
+{
+	$analyzerUrl="https://github.com/Microsoft/dotnet-apiport/releases/download/v1.4.0.alpha.00175/ApiPort.zip"
+	$analyzerZip="$(pwd)\analyzer.zip"
+	wget $analyzerUrl -OutFile $analyzerZip
+	Add-Type -assembly "System.IO.Compression.Filesystem";
+	[IO.Compression.Zipfile]::ExtractToDirectory($analyzerZip, (pwd));
```
Comment (markusweimer) [inline lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1]: we might want to put that into the `bin` folder or such to make sure cleaning up the build folder is easy.

File: lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1 [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+{
+	Remove-Item $outputDirectory -Force -Recurse
+}
+New-Item -ItemType Directory -Force -Path $outputDirectory | Out-Null
+
+# Install the API Portability Analyzer if it is not present
+$portabilityAnalyzer = ".\ApiPort\ApiPort.exe"
+If (!(Test-Path $portabilityAnalyzer))
+{
+	$analyzerUrl="https://github.com/Microsoft/dotnet-apiport/releases/download/v1.4.0.alpha.00175/ApiPort.zip"
```
Comment (markusweimer) [inline lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1]: Any chance to make that a reference to the newest tool? Such that we don't have to update by hand?

File: lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1 [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+			$dotNetTargets = $exeDotNetTargets
+		}
+		Else
+		{
+			echo "Skipping ${projectName}: OutputType is not 'Library','Exe'"
+			return
+		}
+
+		# Define the object we want to explore
+		$objectPath = "${binDir}\${projectName}\${projectName}.${objectSuffix}"
```
Comment (markusweimer) [inline lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1]: This this fail explicitly when the file isn't found?

**Comment by tcNickolas** [comment]
Can we put this script to reef/dev or reef/bin folder? These are the folders where our build and test-related scripts reside, it's convenient to keep them all in one place. Especially if we want to run this script in AppVeyor eventually.

Speaking of running in AppVeyor, can you add it to https://github.com/apache/reef/blob/master/appveyor.yml? There is an `after_build` step in AppVeyor configuration which seems most appropriate. This way it would print the result before running tests, so failed tests wouldn't block compatibility check

**Comment by tcNickolas** [comment]
@markusweimer At a glance this will 1) break in AppVeyor because it always builds from scratch and doesn't keep history, and 2) require that the reviewer runs build for master then build for pull request to get the correct value. There are some CI systems which allow to store information between consecutive builds (I think I've seen one for Cobertura when I was looking at Java tooling) but I don't see an easy way to set it up. Unless we write the result to some remote location accessible from all build workstations which would keep versioning and branch information... which stops being easy

**Comment by markusweimer** [comment]
@tcNickolas, I was thinking of checking in the number of allowed warnings. That way, a PR that would add code that adds new warnings would have to change that line or file as well to pass the tests. And we can then make a decision in the discussion here on whether or not (more likely) we allow it.

**Comment by tcNickolas** [comment]
@markusweimer I like the idea. But it will increase the number of merge conflicts (even if two pull requests change different projects, the total number of warnings can change in both, resulting in a conflict), which AppVeyor doesn't handle well (we don't get merge conflicts often these days but I think it just refuses to do the build). We have to think about this carefully, what other consequences it might have. 

For now I'll just be happy with the script running automatically in AppVeyor and showing the number of warnings.

File: lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1 (hunk 4 above)
Comment (rogancarr) [inline lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1]: Can do! One question though -- the bin folder doesn't clean up when `clean`ed with Visual Studio. Is this expected?

File: lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1 (hunk 5 above)
Comment (rogancarr) [inline lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1]: I agree, but they don't have a URL for this. I opened an issue on the ApiPort project ([ApiPort-449](https://github.com/Microsoft/dotnet-apiport/issues/449)).

File: lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1 (hunk 6 above)
Comment (rogancarr) [inline lang/cs/CoreCLRCompatibility/AnalyzeCoreCLRCompatibility.ps1]: Good eyes. Fixed.

**Comment by rogancarr** [comment]
Here is the latest version. This
* Addresses the code issues that Markus brought up
* Moves it into `\bin`
* Adds an `after_build` step in AppVeyor that runs the script
* Creates all artifacts in the `\lang\cs\bin\x64\Debug\` directory
* Creates a file with the number of incompatibilities in `\lang\cs\bin\x64\Debug\ClrCompatibility\incompatibilities.txt`

Possible Issues:
* This script only targets `Debug` builds.
* Nothing is done with the incompatibilities file.
* The AppVeyor changes _were not tested_

**Comment by markusweimer** [comment]
LGTM will merge.

**Pull request closed without merging**